### ISSUE TITLE
Proxy accumulating SYN_SENT when blocked by Network Policy

### ISSUE THREAD

**Issue by jhrv** [linkerd/linkerd2#6069]
### What is the issue?

When we had a meshed test pod attempt to connect to an application that was blocked by network policy we saw the proxy consume a large number of resources then become non responsive. On further investigation we saw the proxy had a large number of open TCP connections stuck on SYN_SENT. This contrasted with the behavior we expected to see, which ideally would have been that the connections didn't end up being stuck in SYN_SENT.

### Logs, error output, etc

From exec inside the application pod;

```
bash-5.1# netstat -na | grep -c SYN_SENT
8714
```

Here we see the resource consumption (at this point, the proxy was not functioning)

```
$ kubectl top pod --containers=true | grep testrig
testrig-7896d74668-tnsvg           testrig           9m           22Mi
testrig-7896d74668-tnsvg           linkerd-proxy     948m         2178Mi
``` 

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ can initialize the client
√ can query the control plane API

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-ha-checks
-----------------
‼ pod injection disabled on kube-system
    kube-system namespace needs to have the label config.linkerd.io/admission-webhooks: disabled if injector webhook failure policy is Fail
    see https://linkerd.io/checks/#l5d-injection-disabled for hints
√ multiple replicas of control plane pods

Status check results are √

Linkerd extensions checks
=========================

linkerd-viz
-----------
√ linkerd-viz Namespace exists
√ linkerd-viz ClusterRoles exist
√ linkerd-viz ClusterRoleBindings exist
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ tap API service is running
√ linkerd-viz pods are injected
√ viz extension pods are running
√ prometheus is installed and configured correctly
√ can initialize the client
√ viz extension self-check

Status check results are √
```

### Environment

- Kubernetes Version: v1.17.17-gke.3700
- Cluster Environment: GKE
- Host OS: coos (google)
- Linkerd version: 2.10.1

**Comment by adleong** [comment]
Thanks @jhrv!  In order to appropriately prioritize this, it would be helpful to know the impact of this issue.  Do you have a use-case that involves the proxy being blocked by a Network Policy?  Or is this more of a theoretical correctness/hygiene issue?

**Comment by jhrv** [comment]
I would argue it is more than a hygiene/correctness concern. 

We saw this with prometheus bundled with linkerd-viz when migrating, as it was aggressively scraping endpoints which it weren't able to communicate with (yet). Basically any workload with a netpol misconfiguration I believe would suffer from this behavior.

**Comment by olix0r** [comment]
It would be great to get a deeper understanding of what's going on here. I'm removing the 2.11 milestone, as we plan to have an alternate means of policy configuration in this release.

Also, it might be good to test the most recent edge release (edge-21.4.5) which has proxy fixes that will be included in stable-2.10.2 -- we found an issue where the proxy could continually retry failed connections when it should not.
